File: dotnet/Gherkin/AstBuilder.cs
using Gherkin.Ast;

namespace Gherkin;

public class AstBuilder<T> : IAstBuilder<T>
{
    private readonly Stack<AstNode> _stack = new();
    private AstNode CurrentNode => _stack.Peek();
    private readonly List<Comment> _comments = new();

    public AstBuilder()
    {
        Reset();
    }

    public void Reset()
    {
        _stack.Clear();
        _stack.Push(new AstNode(RuleType.None));
        _comments.Clear();
    }

    public void Build(Token token)
    {
        if (token.MatchedType == TokenType.Comment)
        {
            _comments.Add(CreateComment(GetLocation(token), token.MatchedText));
        }
        else
        {
            CurrentNode.Add((RuleType)token.MatchedType, token);
        }
    }

    public void StartRule(RuleType ruleType)
    {
        _stack.Push(new AstNode(ruleType));
    }

    public void EndRule(RuleType ruleType)
    {
        var node = _stack.Pop();
        object transformedNode = GetTransformedNode(node);
        CurrentNode.Add(node.RuleType, transformedNode);
    }

    public T GetResult()
    {
        return CurrentNode.GetSingle<T>(RuleType.GherkinDocument);
    }

    private object GetTransformedNode(AstNode node)
    {
        switch (node.RuleType)
        {
            case RuleType.Step:
                {
                    var stepLine = node.GetToken(TokenType.StepLine);
                    var stepArg = node.GetSingle<StepArgument>(RuleType.DataTable) ??
                                  node.GetSingle<StepArgument>(RuleType.DocString);
                    var keywordType = GetKeywordType(stepLine);
                    return CreateStep(GetLocation(stepLine), stepLine.MatchedKeyword, keywordType, stepLine.MatchedText, stepArg, node);
                }
            case RuleType.DocString:
                {
                    var separatorToken = node.GetTokens(TokenType.DocStringSeparator).First();
                    var contentType = separatorToken.MatchedText.Length == 0 ? null : separatorToken.MatchedText;
                    var lineTokens = node.GetTokens(TokenType.Other);
                    var content = string.Join(Environment.NewLine, lineTokens.Select(lt => lt.MatchedText));
                    var delimiter = separatorToken.MatchedKeyword.Length == 0 ? null : separatorToken.MatchedKeyword;

                    return CreateDocString(GetLocation(separatorToken), contentType, content, delimiter, node);
                }
            case RuleType.DataTable:
                {
                    var rows = GetTableRows(node);
                    return CreateDataTable(rows, node);
                }
            case RuleType.Background:
                {
                    var backgroundLine = node.GetToken(TokenType.BackgroundLine);
                    var description = GetDescription(node);
                    var steps = GetSteps(node);
                    return CreateBackground(GetLocation(backgroundLine), backgroundLine.MatchedKeyword, backgroundLine.MatchedText, description, steps, node);
                }
            case RuleType.ScenarioDefinition:
                {
                    var tags = GetTags(node);

                    var scenarioNode = node.GetSingle<AstNode>(RuleType.Scenario);
                    var scenarioLine = scenarioNode.GetToken(TokenType.ScenarioLine);

                    var description = GetDescription(scenarioNode);
                    var steps = GetSteps(scenarioNode);
                    List<Examples> examples = [.. scenarioNode.GetItems<Examples>(RuleType.ExamplesDefinition)];
                    return CreateScenario(tags, GetLocation(scenarioLine), scenarioLine.MatchedKeyword, scenarioLine.MatchedText, description, steps, examples, node);
                }
            case RuleType.ExamplesDefinition:
                {
                    var tags = GetTags(node);
                    var examplesNode = node.GetSingle<AstNode>(RuleType.Examples);
                    var examplesLine = examplesNode.GetToken(TokenType.ExamplesLine);
                    var description = GetDescription(examplesNode);

                    var allRows = examplesNode.GetSingle<List<TableRow>>(RuleType.ExamplesTable);
                    var header = allRows != null ? allRows[0] : null;
                    var rows = allRows != null ? allRows.Skip(1).ToList() : null;
                    return CreateExamples(tags, GetLocation(examplesLine), examplesLine.MatchedKeyword, examplesLine.MatchedText, description, header, rows, node);
                }
            case RuleType.ExamplesTable:
                {
                    return GetTableRows(node);
                }
            case RuleType.Description:
                {
                    IEnumerable<Token> lineTokens = node.GetTokens(TokenType.Other);

                    // Trim trailing empty lines
                    lineTokens = lineTokens.Reverse().SkipWhile(t => string.IsNullOrWhiteSpace(t.MatchedText)).Reverse();

                    return string.Join(Environment.NewLine, lineTokens.Select(lt => lt.MatchedText));
                }
            case RuleType.Feature:
                {
                    var header = node.GetSingle<AstNode>(RuleType.FeatureHeader);
                    if (header == null) return null;
                    var tags = GetTags(header);
                    var featureLine = header.GetToken(TokenType.FeatureLine);
                    if (featureLine == null) return null;
                    var children = new List<IHasLocation>();
                    var background = node.GetSingle<Background>(RuleType.Background);
                    if (background != null)
                        children.Add(background);
                    foreach (var scenarioDefinition in node.GetItems<IHasLocation>(RuleType.ScenarioDefinition))
                        children.Add(scenarioDefinition);
                    foreach (var rule in node.GetItems<IHasLocation>(RuleType.Rule))
                        children.Add(rule);
                    var description = GetDescription(header);
                    if (featureLine.MatchedGherkinDialect == null) return null;
                    var language = featureLine.MatchedGherkinDialect.Language;

                    return CreateFeature(tags, GetLocation(featureLine), language, featureLine.MatchedKeyword, featureLine.MatchedText, description, children, node);
                }
            case RuleType.Rule:
                {
                    var header = node.GetSingle<AstNode>(RuleType.RuleHeader);
                    if (header == null) return null;
                    var tags = GetTags(header);
                    var ruleLine = header.GetToken(TokenType.RuleLine);
                    if (ruleLine == null) return null;
                    var children = new List<IHasLocation>();
                    var background = node.GetSingle<Background>(RuleType.Background);
                    if (background != null)
                        children.Add(background);
                    foreach (var scenarioDefinition in node.GetItems<IHasLocation>(RuleType.ScenarioDefinition))
                        children.Add(scenarioDefinition);
                    var description = GetDescription(header);
                    if (ruleLine.MatchedGherkinDialect == null) return null;

                    return CreateRule(tags, GetLocation(ruleLine), ruleLine.MatchedKeyword, ruleLine.MatchedText, description, children, node);
                }
            case RuleType.GherkinDocument:
                {
                    var feature = node.GetSingle<Feature>(RuleType.Feature);

                    return CreateGherkinDocument(feature, _comments.ToArray(), node);
                }
        }

        return node;
    }

    protected virtual StepKeywordType GetKeywordType(Token stepLine)
    {
        var stepKeywordType = stepLine.MatchedGherkinDialect.GetStepKeywordType(stepLine.MatchedKeyword);
        if (stepKeywordType == null || stepKeywordType == StepKeywordType.Unspecified)
            return StepKeywordType.Unspecified;
        return stepKeywordType.Value;
    }

    protected virtual Background CreateBackground(Location location, string keyword, string name, string description, IEnumerable<Step> steps, AstNode node)
    {
        return new Background(location, keyword, name, description, steps);
    }

    protected virtual DataTable CreateDataTable(List<TableRow> rows, AstNode node)
    {
        return new DataTable(rows);
    }

    protected virtual Comment CreateComment(Location location, string text)
    {
        return new Comment(location, text);
    }

    protected virtual Examples CreateExamples(IEnumerable<Tag> tags, Location location, string keyword, string name, string description, TableRow header, IEnumerable<TableRow> body, AstNode node)
    {
        return new Examples(tags, location, keyword, name, description, header, body);
    }

    protected virtual Scenario CreateScenario(IEnumerable<Tag> tags, Location location, string keyword, string name, string description, IEnumerable<Step> steps, IEnumerable<Examples> examples, AstNode node)
    {
        return new Scenario(tags, location, keyword, name, description, steps, examples);
    }

    protected virtual DocString CreateDocString(Location location, string contentType, string content, string delimiter, AstNode node)
    {
        return new DocString(location, contentType, content, delimiter);
    }

    protected virtual Step CreateStep(Location location, string keyword, StepKeywordType keywordType, string text, StepArgument argument, AstNode node)
    {
        return new Step(location, keyword, keywordType, text, argument);
    }

    protected virtual GherkinDocument CreateGherkinDocument(Feature feature, IEnumerable<Comment> gherkinDocumentComments, AstNode node)
    {
        return new GherkinDocument(feature, gherkinDocumentComments);
    }

    protected virtual Feature CreateFeature(IEnumerable<Tag> tags, Location location, string language, string keyword, string name, string description, IEnumerable<IHasLocation> children, AstNode node)
    {
        return new Feature(tags, location, language, keyword, name, description, children);
    }

    protected virtual Rule CreateRule(IEnumerable<Tag> tags, Location location, string keyword, string name, string description, IEnumerable<IHasLocation> children, AstNode node)
    {
        return new Rule(tags, location, keyword, name, description, children);
    }

    protected virtual Tag CreateTag(Location location, string name, AstNode node)
    {
        return new Tag(location, name);
    }

    protected Location CreateLocation(int line, int column)
    {
        return new Location(line, column);
    }

    protected virtual TableRow CreateTableRow(Location location, IEnumerable<TableCell> cells, AstNode node)
    {
        return new TableRow(location, cells);
    }

    protected TableCell CreateTableCell(Location location, string value)
    {
        return new TableCell(location, value);
    }

    private Location GetLocation(Token token, int column = 0)
    {
        return column == 0 ? token.Location : CreateLocation(token.Location.Line, column);
    }

    private IEnumerable<Tag> GetTags(AstNode node)
    {
        var tagsNode = node.GetSingle<AstNode>(RuleType.Tags);
        if (tagsNode == null)
            return [];

        var tags = new List<Tag>();
        foreach (var line in tagsNode.GetTokens(TokenType.TagLine))
        {
            foreach (var matchedItem in line.MatchedItems)
                tags.Add(CreateTag(GetLocation(line, matchedItem.Column), matchedItem.Text, tagsNode));
        }
        return tags;
    }

    private List<TableRow> GetTableRows(AstNode node)
    {
        var rows = new List<TableRow>();
        int cellCount = 0;
        bool firstRow = true;
        foreach (var rowToken in node.GetTokens(TokenType.TableRow))
        {
            var rowLocation = GetLocation(rowToken);
            var cells = new List<TableCell>();
            foreach (var cellItem in rowToken.MatchedItems)
                cells.Add(CreateTableCell(GetLocation(rowToken, cellItem.Column), cellItem.Text));
            if (firstRow)
            {
                cellCount = cells.Count;
                firstRow = false;
            }
            else if (cells.Count != cellCount)
            {
                HandleAstError("inconsistent cell count within the table", rowLocation);
            }
            rows.Add(CreateTableRow(rowLocation, cells, node));
        }
        return rows;
    }

    protected virtual void HandleAstError(string message, Location location)
    {
        throw new AstBuilderException(message, location);
    }

    private static List<Step> GetSteps(AstNode scenarioDefinitionNode)
    {
        return [..scenarioDefinitionNode.GetItems<Step>(RuleType.Step)];
    }

    private static string GetDescription(AstNode scenarioDefinitionNode)
    {
        return scenarioDefinitionNode.GetSingle<string>(RuleType.Description);
    }
}

File: python/src/gherkin/ast_builder.py
from __future__ import annotations

from typing import TypedDict, cast, TypeVar, Union

from .ast_node import AstNode
from .errors import AstBuilderException
from .parser_types import (
    GherkinDocument,
    Tag,
    TableRow,
    Cell,
    Step,
    DocString,
    Scenario,
    Background,
    DataTable,
    Examples,
    Rule,
    Feature,
)
from .stream.id_generator import IdGenerator
from .token import Location, Token

T = TypeVar("T")


class Comment(TypedDict):
    location: Location
    text: str


class AstBuilder:
    id_generator: IdGenerator
    stack: list[AstNode]
    comments: list[Comment]
    id_counter: int

    def __init__(self, id_generator: IdGenerator | None = None) -> None:
        if id_generator is None:
            id_generator = IdGenerator()
        self.id_generator = id_generator
        self.reset()

    def reset(self) -> None:
        self.stack = [AstNode("None")]
        self.comments = []
        self.id_counter = 0

    def start_rule(self, rule_type: str) -> None:
        self.stack.append(AstNode(rule_type))

    def end_rule(self, rule_type: str) -> None:
        node = self.stack.pop()
        self.current_node.add(node.rule_type, self.transform_node(node))

    def build(self, token: Token) -> None:
        if token.matched_type == "Comment":
            self.comments.append(
                {"location": self.get_location(token), "text": token.matched_text}
            )
        else:
            self.current_node.add(token.matched_type, token)

    def get_result(self) -> object:
        return self.current_node.get_single("GherkinDocument")

    @property
    def current_node(self) -> AstNode:
        return self.stack[-1]

    @staticmethod
    def get_location(token: Token, column: int | None = None) -> Location:
        return (
            token.location
            if not column
            else {"line": token.location["line"], "column": column}
        )

    def get_tags(self, node: AstNode) -> list[Tag]:
        tags: list[Tag] = []
        tags_node = cast(AstNode, node.get_single("Tags"))
        if not tags_node:
            return tags

        for token in tags_node.get_tokens("TagLine"):
            tags += [
                {
                    "id": self.id_generator.get_next_id(),
                    "location": self.get_location(token, tag_item["column"]),
                    "name": tag_item["text"],
                }
                for tag_item in token.matched_items
            ]

        return tags

    def get_table_rows(self, node: AstNode) -> list[TableRow]:
        rows: list[TableRow] = [
            {
                "id": self.id_generator.get_next_id(),
                "location": self.get_location(token),
                "cells": self.get_cells(token),
            }
            for token in node.get_tokens("TableRow")
        ]
        self.ensure_cell_count(rows)
        return rows

    @staticmethod
    def ensure_cell_count(rows: list[TableRow]) -> None:
        if not rows:
            return

        cell_count = len(rows[0]["cells"])
        for row in rows:
            if len(row["cells"]) != cell_count:
                raise AstBuilderException(
                    "inconsistent cell count within the table", row["location"]
                )

    def get_cells(self, table_row_token: Token) -> list[Cell]:
        return [
            cast(
                Cell,
                self.reject_nones(
                    {
                        "location": self.get_location(
                            table_row_token, cell_item["column"]
                        ),
                        "value": cell_item["text"],
                    }
                ),
            )
            for cell_item in table_row_token.matched_items
        ]

    @staticmethod
    def get_description(node: AstNode) -> str:
        return cast(str, node.get_single("Description", ""))

    @staticmethod
    def get_steps(node: AstNode) -> list[Step]:
        return cast(list[Step], node.get_items("Step"))

    def transform_node(
        self, node: AstNode
    ) -> (
        Step
        | DocString
        | DataTable
        | Background
        | Scenario
        | Examples
        | list[TableRow]
        | str
        | None
        | Rule
        | Feature
        | GherkinDocument
        | AstNode
    ):
        if node.rule_type == "Step":
            step_line = node.get_token("StepLine")
            step_argument_type = "dummy_type"
            step_argument = None
            if node.get_single("DataTable"):
                step_argument_type = "dataTable"
                step_argument = node.get_single("DataTable")
            elif node.get_single("DocString"):
                step_argument_type = "docString"
                step_argument = node.get_single("DocString")

            return self.reject_nones(
                {
                    "id": self.id_generator.get_next_id(),
                    "location": self.get_location(step_line),
                    "keyword": step_line.matched_keyword,
                    "keywordType": step_line.matched_keyword_type,
                    "text": step_line.matched_text,
                    step_argument_type: step_argument,
                }
            )
        elif node.rule_type == "DocString":
            separator_token = node.get_tokens("DocStringSeparator")[0]
            media_type = (
                separator_token.matched_text
                if len(separator_token.matched_text) > 0
                else None
            )
            line_tokens = node.get_tokens("Other")
            content = "\n".join([t.matched_text for t in line_tokens])

            return self.reject_nones(
                {
                    "location": self.get_location(separator_token),
                    "content": content,
                    "delimiter": separator_token.matched_keyword,
                    "mediaType": media_type,
                }
            )
        elif node.rule_type == "DataTable":
            rows = self.get_table_rows(node)
            return self.reject_nones(
                {
                    "location": rows[0]["location"],
                    "rows": rows,
                }
            )
        elif node.rule_type == "Background":
            background_line = node.get_token("BackgroundLine")
            description = self.get_description(node)
            steps = self.get_steps(node)

            return self.reject_nones(
                {
                    "id": self.id_generator.get_next_id(),
                    "location": self.get_location(background_line),
                    "keyword": background_line.matched_keyword,
                    "name": background_line.matched_text,
                    "description": description,
                    "steps": steps,
                }
            )
        elif node.rule_type == "ScenarioDefinition":
            tags = self.get_tags(node)
            scenario_node = cast(AstNode, node.get_single("Scenario"))
            scenario_line = scenario_node.get_token("ScenarioLine")
            description = self.get_description(scenario_node)
            steps = self.get_steps(scenario_node)
            examples = scenario_node.get_items("ExamplesDefinition")

            return self.reject_nones(
                {
                    "id": self.id_generator.get_next_id(),
                    "tags": tags,
                    "location": self.get_location(scenario_line),
                    "keyword": scenario_line.matched_keyword,
                    "name": scenario_line.matched_text,
                    "description": description,
                    "steps": steps,
                    "examples": examples,
                }
            )
        elif node.rule_type == "ExamplesDefinition":
            tags = self.get_tags(node)
            examples_node = cast(AstNode, node.get_single("Examples"))
            examples_line = examples_node.get_token("ExamplesLine")
            description = self.get_description(examples_node)
            examples_table_rows = cast(
                list[TableRow], examples_node.get_single("ExamplesTable")
            )
            table_header = examples_table_rows[0] if examples_table_rows else None
            table_body = examples_table_rows[1:] if examples_table_rows else []

            return self.reject_nones(
                {
                    "id": self.id_generator.get_next_id(),
                    "tags": tags,
                    "location": self.get_location(examples_line),
                    "keyword": examples_line.matched_keyword,
                    "name": examples_line.matched_text,
                    "description": description,
                    "tableHeader": table_header,
                    "tableBody": table_body,
                }
            )
        elif node.rule_type == "ExamplesTable":
            return self.get_table_rows(node)
        elif node.rule_type == "Description":
            line_tokens = node.get_tokens("Other")
            tokens = list(line_tokens)

            # Trim trailing empty lines
            while tokens and not tokens[-1].matched_text:
                tokens.pop()

            return "\n".join(token.matched_text for token in tokens)
        elif node.rule_type == "Rule":
            header = cast(Union[AstNode, None], node.get_single("RuleHeader"))
            if not header:
                return None
            tags = self.get_tags(header)
            rule_line = header.get_token("RuleLine")
            if not rule_line:
                return None

            children = []
            background = node.get_single("Background")
            if background:
                children.append({"background": background})
            children = children + [
                {"scenario": i} for i in node.get_items("ScenarioDefinition")
            ]
            description = self.get_description(header)

            return self.reject_nones(
                {
                    "id": self.id_generator.get_next_id(),
                    "tags": tags,
                    "location": self.get_location(rule_line),
                    "keyword": rule_line.matched_keyword,
                    "name": rule_line.matched_text,
                    "description": description,
                    "children": children,
                }
            )
        elif node.rule_type == "Feature":
            header = cast(Union[AstNode, None], node.get_single("FeatureHeader"))
            if not header:
                return None

            tags = self.get_tags(header)
            feature_line = header.get_token("FeatureLine")
            if not feature_line:
                return None

            children = []
            background = node.get_single("Background")
            if background:
                children.append({"background": background})
            children = children + [
                {"scenario": i} for i in node.get_items("ScenarioDefinition")
            ]
            children = children + [{"rule": i} for i in node.get_items("Rule")]
            description = self.get_description(header)
            language = feature_line.matched_gherkin_dialect

            return self.reject_nones(
                {
                    "tags": tags,
                    "location": self.get_location(feature_line),
                    "language": language,
                    "keyword": feature_line.matched_keyword,
                    "name": feature_line.matched_text,
                    "description": description,
                    "children": children,
                }
            )
        elif node.rule_type == "GherkinDocument":
            feature = node.get_single("Feature")

            return self.reject_nones({"feature": feature, "comments": self.comments})
        else:
            return node

    @staticmethod
    def reject_nones(values: dict[str, T | None]) -> dict[str, T]:
        return {k: v for k, v in values.items() if v is not None}
